File: packages/native/runtime-swift/Sources/PyreonRuntime/PyreonFilePicker.swift
// PyreonFilePicker — the SwiftUI side of Pyreon's document-picker service
// (M3.8). Mirrors the core `@pyreon/hooks` `useFilePicker` shape and the Kotlin
// `PyreonFilePicker` one-for-one. The document sibling of PyreonImagePicker
// (any file, not just photos).
//
// Surface:
//
//     let uri = await files.pick()   // "file:///…" or nil when cancelled
//
// `pick()` is ASYNC — the third Pyreon service with an async RESULT (after
// PyreonBiometrics and PyreonImagePicker). PMTC lowers `const uri = await
// files.pick()` inside an `async` handler to `Task { let uri = await
// files.pick() }` (M4.5). It never throws to the caller (every failure
// collapses to `nil`), so the emitted `await` needs no `try`.
//
// NO storage permission / entitlement is required: UIDocumentPickerViewController
// runs OUT OF PROCESS, so the app never gains broad filesystem access — it
// receives only the single document the user explicitly picked (as a
// security-scoped URL).
//
// UIKit / UniformTypeIdentifiers are guarded with `#if canImport` so the type
// still compiles on platforms without them (the emit references the type shape
// unconditionally). macOS has UniformTypeIdentifiers but not
// UIDocumentPickerViewController, so the whole implementation is gated on UIKit.

import Foundation
#if canImport(UIKit)
import UIKit
import UniformTypeIdentifiers
#endif

public struct PyreonFilePicker {
    public init() {}

    /// Present the system document picker and return the picked file's URI, or
    /// `nil` if the user cancelled (or no picker could be presented).
    ///
    /// The returned string is a `file://` URL pointing at a COPY of the picked
    /// document in the app's temporary directory. The copy is MANDATORY: the
    /// picker hands back a SECURITY-SCOPED URL that is only accessible between
    /// `startAccessingSecurityScopedResource()` / `stop…`, so the bytes must be
    /// copied into an app-owned location before that access ends. Treat the
    /// result as ephemeral: the OS may reclaim `NSTemporaryDirectory()`.
    public func pick() async -> String? {
        #if canImport(UIKit)
        guard let presenter = await PyreonFilePicker.topViewController() else {
            return nil
        }
        return await withCheckedContinuation { (continuation: CheckedContinuation<String?, Never>) in
            Task { @MainActor in
                // `.item` is the broadest UTType — any file. A typed filter is a
                // follow-up (the emit passes no argument today).
                let picker = UIDocumentPickerViewController(forOpeningContentTypes: [UTType.item])
                picker.allowsMultipleSelection = false

                // The continuation MUST resume exactly once. didPick and
                // wasCancelled are mutually exclusive, but guard anyway — a
                // double-resume traps.
                let resumed = PyreonResumeGuard()
                let delegate = PyreonDocumentPickerDelegate(
                    onPick: { url in
                        if resumed.claim() {
                            continuation.resume(returning: PyreonFilePicker.copyToTemporaryFile(url))
                        }
                    },
                    onCancel: {
                        if resumed.claim() { continuation.resume(returning: nil) }
                    }
                )
                // UIDocumentPickerViewController holds its delegate WEAKLY, so
                // the delegate must outlive this scope or the continuation is
                // never resumed and the caller's `await` hangs forever. Park it
                // on the picker so it lives exactly as long as the presentation.
                picker.delegate = delegate
                objc_setAssociatedObject(
                    picker,
                    &PyreonFilePicker.delegateKey,
                    delegate,
                    .OBJC_ASSOCIATION_RETAIN_NONATOMIC
                )
                presenter.present(picker, animated: true)
            }
        }
        #else
        return nil
        #endif
    }

    #if canImport(UIKit)
    private static var delegateKey: UInt8 = 0

    /// The view controller to present from — the key window's root, walked down
    /// through any already-presented controllers so we never try to present on
    /// top of a controller that is itself presenting (which UIKit rejects).
    @MainActor
    private static func topViewController() -> UIViewController? {
        let scenes = UIApplication.shared.connectedScenes
        let windowScene = scenes.compactMap { $0 as? UIWindowScene }.first { $0.activationState == .foregroundActive }
            ?? scenes.compactMap { $0 as? UIWindowScene }.first
        guard var top = windowScene?.windows.first(where: { $0.isKeyWindow })?.rootViewController
            ?? windowScene?.windows.first?.rootViewController
        else { return nil }
        while let presented = top.presentedViewController {
            top = presented
        }
        return top
    }

    /// Copy the picked document's bytes into the temp directory and return a
    /// `file://` URL string. The source URL is security-scoped, so access must
    /// be bracketed by `start/stopAccessingSecurityScopedResource()`.
    fileprivate static func copyToTemporaryFile(_ url: URL) -> String? {
        let didAccess = url.startAccessingSecurityScopedResource()
        defer { if didAccess { url.stopAccessingSecurityScopedResource() } }
        let destination = URL(fileURLWithPath: NSTemporaryDirectory())
            .appendingPathComponent("pyreon-\(UUID().uuidString)-\(url.lastPathComponent)")
        do {
            try FileManager.default.copyItem(at: url, to: destination)
            return destination.absoluteString
        } catch {
            return nil
        }
    }
    #endif
}

#if canImport(UIKit)
/// One-shot claim flag guarding a CheckedContinuation against a double-resume.
///
/// `@unchecked Sendable` is accurate rather than a silencer: the only mutable
/// state (`used`) is read and written exclusively under `lock`, so crossing the
/// `@Sendable` boundary is safe.
private final class PyreonResumeGuard: @unchecked Sendable {
    private let lock = NSLock()
    private var used = false
    func claim() -> Bool {
        lock.lock()
        defer { lock.unlock() }
        if used { return false }
        used = true
        return true
    }
}

/// Bridges UIDocumentPickerViewController's delegate callbacks to the async
/// continuation. UIDocumentPickerViewController auto-dismisses on pick/cancel,
/// so — unlike PHPicker — this delegate does not dismiss it manually.
private final class PyreonDocumentPickerDelegate: NSObject, UIDocumentPickerDelegate {
    private let onPick: (URL) -> Void
    private let onCancel: () -> Void

    init(onPick: @escaping (URL) -> Void, onCancel: @escaping () -> Void) {
        self.onPick = onPick
        self.onCancel = onCancel
    }

    func documentPicker(_ controller: UIDocumentPickerViewController, didPickDocumentsAt urls: [URL]) {
        guard let url = urls.first else {
            onCancel()
            return
        }
        onPick(url)
    }

    func documentPickerWasCancelled(_ controller: UIDocumentPickerViewController) {
        onCancel()
    }
}
#endif

File: packages/native/runtime-kotlin/src/main/kotlin/com/pyreon/runtime/PyreonFilePicker.kt
// PyreonFilePicker — the Compose side of Pyreon's document-picker service
// (M3.8). Mirrors the Swift `PyreonFilePicker` one-for-one + the core
// `@pyreon/hooks` `useFilePicker` shape. The document sibling of
// PyreonImagePicker (any file, not just photos).
//
// Surface:
//
//     val uri = files.pick()   // "content://…" or null when cancelled
//
// `pick()` is a `suspend` fun — the third Pyreon service with an async RESULT
// (after PyreonBiometrics and PyreonImagePicker). PMTC lowers `const uri = await
// files.pick()` inside an `async` handler to `pyreonAsyncScope.launch { val uri
// = files.pick() }` (the M4.5 lowering; Kotlin suspend calls carry no `await`).
//
// NO storage permission is required: the Storage Access Framework
// (`OpenDocument`) runs OUT OF PROCESS and hands back only the single document
// the user picked — the same policy-safe property as iOS's
// UIDocumentPickerViewController.
//
// THE iOS/ANDROID ASYMMETRY this class absorbs (identical to PyreonImagePicker):
// iOS presents from the key window, so `PyreonFilePicker.swift` is
// self-contained. Android delivers the result through an ActivityResult CALLBACK
// whose registration must happen at composition time
// (`rememberLauncherForActivityResult` is a @Composable; registering once the
// host is RESUMED throws). So the emit wires a composable-scope launcher into
// `launcher`, and this class bridges callback→suspend via a CompletableDeferred
// — which lets `pick()` keep the exact `suspend fun (): String?` shape as
// Swift's `async`.
//
// The one delta from the image picker: `OpenDocument`'s launcher input is an
// `Array<String>` of MIME types (vs `PickVisualMediaRequest`). `pick()` passes
// `arrayOf("*/*")` — accept any document; a typed filter is a follow-up.
//
// Deps: kotlinx-coroutines + androidx.activity (both already present in every
// Android example that compiles this srcDir).
//
// Threading: `pending` is touched from `pick()` (a coroutine started by
// `rememberCoroutineScope()`, whose context is Main) and from `onResult` (an
// ActivityResult callback, also Main), so both touches are main-thread-confined
// and need no synchronisation.

package com.pyreon.runtime

import androidx.activity.result.ActivityResultLauncher
import kotlinx.coroutines.CompletableDeferred

class PyreonFilePicker {
    /**
     * The composable-scope ActivityResult launcher, assigned by the PMTC emit
     * (`files.launcher = rememberLauncherForActivityResult(OpenDocument()) { … }`).
     * Its input is the `Array<String>` of acceptable MIME types the SAF
     * `OpenDocument` contract takes. Null until composition wires it — `pick()`
     * then resolves null rather than throwing, so a caller's `await` can never
     * hang.
     */
    var launcher: ActivityResultLauncher<Array<String>>? = null

    private var pending: CompletableDeferred<String?>? = null

    /**
     * Deliver an ActivityResult back to the suspended `pick()`. Called by the
     * emit-wired launcher callback with the picked `Uri.toString()`, or null
     * when the user cancelled.
     */
    fun onResult(uri: String?) {
        val deferred = pending
        pending = null
        deferred?.complete(uri)
    }

    /**
     * Present the system document picker and return the picked file's
     * `content://` URI, or null if the user cancelled (or no launcher is
     * wired). Never throws.
     */
    suspend fun pick(): String? {
        val activeLauncher = launcher ?: return null
        // A second pick() while one is still in flight: settle the older one as
        // cancelled so its caller's await can never hang forever (only one
        // ActivityResult will come back, and it belongs to the newer launch).
        pending?.complete(null)
        val deferred = CompletableDeferred<String?>()
        pending = deferred
        // "*/*" — accept any document type. A typed filter is a follow-up.
        activeLauncher.launch(arrayOf("*/*"))
        return deferred.await()
    }
}
